handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -917,16 +927,19 @@ public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
         discardSomeReadBytes();
 
         flushIfNeeded(ctx);
+        readIfNeeded(ctx);
+
+        firedChannelRead = false;
+        ctx.fireChannelReadComplete();
+    }
 
+    private void readIfNeeded(ChannelHandlerContext ctx) {
         // If handshake is not finished yet, we need more data.
         if (!ctx.channel().config().isAutoRead() && (!firedChannelRead || !handshakePromise.isDone())) {
             // No auto-read used and no message passed through the ChannelPipeline or the handhshake was not complete
             // yet, which means we need to trigger the read to ensure we not encounter any stalls.
             ctx.read();
```

Comment:
One more thing. I'm not 100% sure we need to change `readDuringHandshake` via `readIfNeeded` since it serves a different propose. I'm don't think we should count the read `SslHandler` initiated to handle the handshake as `readDuringHandshake`. Otherwise it should be `true` all the the time. Although, I'm pretty sure I'm missing something here.